### PR TITLE
lgc/DescBuilder: assert that resource nodes can be found

### PR DESCRIPTION
**Pull request by AMD-dwang**
These assertions should always be true, and if they did fail, then continuing with an `undef` value isn't much better because it will result in incorrect code anyway.

As nice side effect, this happens to likely paper over some issues with an internal use case of the compiler.

Add the matched resource for some tests.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit db256396ebf1021e05f68e1cffd306e8766f23b2
<details><summary>CTS tests (Failed: 2/286511)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.6, Gfx8</summary><ul>

<li>Passed: 36222/72648 (49.9%)</li>
<li>Failed: 0/72648 (0.0%)</li>
<li>Not Supported: 36426/72648 (50.1%)</li><li>Warnings: 0/72648 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi21</summary><ul>

<li>Passed: 40418/72695 (55.6%)</li>
<li><details><summary>Failed: 2/72695 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.binding_model.descriptorset_random.sets8.dynindexed.ubolimitlow.sbolimitlow.sampledimglow.lowimgnotex.noiub.uab.sect.noia.0
Stack trace: Flake

FAILURE: dEQP-VK.ray_tracing_pipeline.builtin.launchidext.sect_317_3331_1
Stack trace: Flake
```
</p></details></li>

<li>Not Supported: 32275/72695 (44.4%)</li><li>Warnings: 0/72695 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 37135/72702 (51.1%)</li>
<li>Failed: 0/72702 (0.0%)</li>
<li>Not Supported: 35567/72702 (48.9%)</li><li>Warnings: 0/72702 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx9</summary><ul>

<li>Passed: 34834/68466 (50.9%)</li>
<li>Failed: 0/68466 (0.0%)</li>
<li>Not Supported: 33632/68466 (49.1%)</li><li>Warnings: 0/68466 (0.0%)</li>
</ul></details>


</ul></details>

File: lgc/builder/DescBuilder.cpp [hunk 1]
```diff
@@ -164,10 +161,11 @@ Value *DescBuilder::CreateGetDescStride(ResourceNodeType concreteType, ResourceN
...
   if (!m_pipelineState->isUnlinked() || !m_pipelineState->getUserDataNodes().empty()) {
     std::tie(topNode, node) = m_pipelineState->findResourceNode(abstractType, descSet, binding);
-    if (!node) {
-      // We did not find the resource node. Return an undef value.
+    // NOTE: Resource node may be DescriptorTexelBuffer, but it is defined as OpTypeSampledImage in SPIRV,
+    // The unused DescriptorSampler may need to be found, in this case we will return undefined value.
+    if (!node && m_pipelineState->findResourceNode(ResourceNodeType::Unknown, descSet, binding).second)
       return UndefValue::get(getInt32Ty());
-    }
+    assert(node && "missing resource node");
```
Comment (nhaehnle) [inline lgc/builder/DescBuilder.cpp:168]: Can we just return `nullptr` here instead of `undef` and have the caller check for that? `nullptr` is a clearer, easier-to-check-for, and harder-to-accidentally-ignore indication of "not found". That is:
```
    // NOTE: Resource node may be DescriptorTexelBuffer, but it is defined as OpTypeSampledImage in SPIRV,
    // In this case, a caller may search for the DescriptorSampler and not find it. We return nullptr in this case and
    // and expect the caller to handle it.
    return node;
```
And the same also below.

This should require some changes elsewhere, but it would be overall cleaner.

**Review by nhaehnle — COMMENTED** [review]
Thank you for looking into this. This is indeed a better solution but I think it could still be improved.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 017bb370e8aa7139dd94566e9bf35fcacd84e008
<details><summary>CTS tests (Failed: 53/286514)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.6, Gfx8</summary><ul>

<li>Passed: 36210/72648 (49.8%)</li>
<li><details><summary>Failed: 12/72648 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.copy_and_blit.copy_commands2.resolve_image.whole_array_image.2_bit
Stack trace: Crash

FAILURE: dEQP-VK.api.copy_and_blit.copy_commands2.resolve_image.whole_array_image.4_bit
Stack trace: Crash

FAILURE: dEQP-VK.api.copy_and_blit.copy_commands2.resolve_image.whole_array_image.8_bit
Stack trace: Crash

FAILURE: dEQP-VK.api.copy_and_blit.copy_commands2.resolve_image.whole_array_image_one_region.2_bit
Stack trace: Crash

FAILURE: dEQP-VK.api.copy_and_blit.copy_commands2.resolve_image.whole_array_image_one_region.8_bit
Stack trace: Crash

FAILURE: dEQP-VK.api.copy_and_blit.core.resolve_image.whole_array_image.8_bit
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.core.resolve_image.whole_array_image_one_region.2_bit
Stack trace: Crash

FAILURE: dEQP-VK.api.copy_and_blit.core.resolve_image.whole_array_image_one_region.8_bit
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.resolve_image.whole_array_image.2_bit
Stack trace: Crash

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.resolve_image.whole_array_image.4_bit
Stack trace: Crash

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.resolve_image.whole_array_image.8_bit
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.resolve_image.whole_array_image_one_region.8_bit
Stack trace: Crash
```
</p></details></li>

<li>Not Supported: 36426/72648 (50.1%)</li><li>Warnings: 0/72648 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi21</summary><ul>

<li>Passed: 40408/72695 (55.6%)</li>
<li><details><summary>Failed: 12/72695 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.copy_and_blit.copy_commands2.resolve_image.whole_array_image.2_bit
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.copy_commands2.resolve_image.whole_array_image.8_bit
Stack trace: Crash

FAILURE: dEQP-VK.api.copy_and_blit.copy_commands2.resolve_image.whole_array_image_one_region.4_bit
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.copy_commands2.resolve_image.whole_array_image_one_region.8_bit
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.core.resolve_image.whole_array_image.2_bit
Stack trace: Crash

FAILURE: dEQP-VK.api.copy_and_blit.core.resolve_image.whole_array_image.4_bit
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.core.resolve_image.whole_array_image.8_bit
Stack trace: Crash

FAILURE: dEQP-VK.api.copy_and_blit.core.resolve_image.whole_array_image_one_region.4_bit
Stack trace: Crash

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.resolve_image.whole_array_image.2_bit
Stack trace: Crash

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.resolve_image.whole_array_image_one_region.2_bit
Stack trace: Crash

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.resolve_image.whole_array_image_one_region.4_bit
Stack trace: Crash

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.resolve_image.whole_array_image_one_region.8_bit
Stack trace: Crash
```
</p></details></li>

<li>Not Supported: 32275/72695 (44.4%)</li><li>Warnings: 0/72695 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 37119/72702 (51.1%)</li>
<li><details><summary>Failed: 16/72702 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.copy_and_blit.copy_commands2.resolve_image.whole_array_image.2_bit
Stack trace: Crash

FAILURE: dEQP-VK.api.copy_and_blit.copy_commands2.resolve_image.whole_array_image.4_bit
Stack trace: Crash

FAILURE: dEQP-VK.api.copy_and_blit.copy_commands2.resolve_image.whole_array_image.8_bit
Stack trace: Crash

FAILURE: dEQP-VK.api.copy_and_blit.copy_commands2.resolve_image.whole_array_image_one_region.2_bit
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.copy_commands2.resolve_image.whole_array_image_one_region.4_bit
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.copy_commands2.resolve_image.whole_array_image_one_region.8_bit
Stack trace: Crash

FAILURE: dEQP-VK.api.copy_and_blit.core.resolve_image.whole_array_image.2_bit
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.core.resolve_image.whole_array_image.4_bit
Stack trace: Crash

FAILURE: dEQP-VK.api.copy_and_blit.core.resolve_image.whole_array_image_one_region.2_bit
Stack trace: Crash

FAILURE: dEQP-VK.api.copy_and_blit.core.resolve_image.whole_array_image_one_region.4_bit
Stack trace: Crash

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.resolve_image.whole_array_image.2_bit
Stack trace: Crash

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.resolve_image.whole_array_image.4_bit
Stack trace: Crash

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.resolve_image.whole_array_image.8_bit
Stack trace: Crash

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.resolve_image.whole_array_image_one_region.2_bit
Stack trace: Crash

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.resolve_image.whole_array_image_one_region.4_bit
Stack trace: Crash

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.resolve_image.whole_array_image_one_region.8_bit
Stack trace: Crash
```
</p></details></li>

<li>Not Supported: 35567/72702 (48.9%)</li><li>Warnings: 0/72702 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx9</summary><ul>

<li>Passed: 34821/68469 (50.9%)</li>
<li><details><summary>Failed: 13/68469 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.copy_and_blit.copy_commands2.resolve_image.whole_array_image.4_bit
Stack trace: Crash

FAILURE: dEQP-VK.api.copy_and_blit.copy_commands2.resolve_image.whole_array_image.8_bit
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.copy_commands2.resolve_image.whole_array_image_one_region.2_bit
Stack trace: Crash

FAILURE: dEQP-VK.api.copy_and_blit.copy_commands2.resolve_image.whole_array_image_one_region.4_bit
Stack trace: Crash

FAILURE: dEQP-VK.api.copy_and_blit.copy_commands2.resolve_image.whole_array_image_one_region.8_bit
Stack trace: Crash

FAILURE: dEQP-VK.api.copy_and_blit.core.resolve_image.whole_array_image.2_bit
Stack trace: Crash

FAILURE: dEQP-VK.api.copy_and_blit.core.resolve_image.whole_array_image.4_bit
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.core.resolve_image.whole_array_image_one_region.4_bit
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.resolve_image.whole_array_image.2_bit
Stack trace: Crash

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.resolve_image.whole_array_image.4_bit
Stack trace: Crash

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.resolve_image.whole_array_image_one_region.2_bit
Stack trace: Crash

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.resolve_image.whole_array_image_one_region.4_bit
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.resolve_image.whole_array_image_one_region.8_bit
Stack trace: Crash
```
</p></details></li>

<li>Not Supported: 33635/68469 (49.1%)</li><li>Warnings: 0/68469 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 6b693762e5bf6685e862c77e1c734f9f72e20934
<details><summary>CTS tests (Failed: 0/286514)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.6, Gfx8</summary><ul>

<li>Passed: 36222/72648 (49.9%)</li>
<li>Failed: 0/72648 (0.0%)</li>
<li>Not Supported: 36426/72648 (50.1%)</li><li>Warnings: 0/72648 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi21</summary><ul>

<li>Passed: 40420/72695 (55.6%)</li>
<li>Failed: 0/72695 (0.0%)</li>
<li>Not Supported: 32275/72695 (44.4%)</li><li>Warnings: 0/72695 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 37135/72702 (51.1%)</li>
<li>Failed: 0/72702 (0.0%)</li>
<li>Not Supported: 35567/72702 (48.9%)</li><li>Warnings: 0/72702 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx9</summary><ul>

<li>Passed: 34834/68469 (50.9%)</li>
<li>Failed: 0/68469 (0.0%)</li>
<li>Not Supported: 33635/68469 (49.1%)</li><li>Warnings: 0/68469 (0.0%)</li>
</ul></details>


</ul></details>

File: lgc/builder/BuilderReplayer.cpp [hunk 2]
```diff
@@ -197,15 +197,15 @@ void BuilderReplayer::replayCall(unsigned opcode, CallInst *call) {
   // Process the builder call.
   LLVM_DEBUG(dbgs() << "Replaying " << *call << "\n");
   Value *newValue = processCall(opcode, call);
+  if (!newValue)
+    newValue = UndefValue::get(call->getType());
```
Comment (nhaehnle) [inline lgc/builder/BuilderReplayer.cpp:201]: Why is that necessary?

Comment (AMD-dwang) [reply lgc/builder/BuilderReplayer.cpp:201]: This function is called in below code
```
 while (!func.use_empty()) {
      CallInst *call = cast<CallInst>(func.use_begin()->getUser());
      replayCall(opcode, call);
    }
```
So we must erase the `call` instruction, otherwise we will entry a endless. but we must replace this `call` before erasing, so undefined value is necessary.

File: lgc/builder/DescBuilder.cpp [hunk 3]
```diff
@@ -164,10 +161,13 @@ Value *DescBuilder::CreateGetDescStride(ResourceNodeType concreteType, ResourceN
...
-    if (!node) {
-      // We did not find the resource node. Return an undef value.
-      return UndefValue::get(getInt32Ty());
+    if (!node && m_pipelineState->findResourceNode(ResourceNodeType::Unknown, descSet, binding).second) {
+      // NOTE: Resource node may be DescriptorTexelBuffer, but it is defined as OpTypeSampledImage in SPIRV,
+      // In this case, a caller may search for the DescriptorSampler and not find it. We return nullptr and
+      // expect the caller to handle it.
+      return nullptr;
     }
+    assert(node && "missing resource node");
```
[the remark this answers is not in the document]
Comment (AMD-dwang) [reply lgc/builder/DescBuilder.cpp:170]: It is not okay. It can't determine if the resource is missing in release version.

File: lgc/builder/BuilderReplayer.cpp (hunk 2 above)
Comment (nhaehnle) [reply lgc/builder/BuilderReplayer.cpp:201]: Hmm, does this mean that there is some other code that expects the replay to be successful and checks explicitly for an `undef` value? That's unfortunate and should really be fixed, because the practice of assigning specific semantic meaning to `undef` doesn't mesh well with how LLVM IR defined and should be avoided, and we should not expand the behavior of relying on `undef` in this way.

Is it easy to fix the larger issue here? If yes, that would be great, but if it isn't then feel free to go back to the previous version where `undef` was created by `findResourceNode`.

Do you have e.g. a CTS test that is affected by this problem and needs the `undef`?

Comment (AMD-dwang) [reply lgc/builder/BuilderReplayer.cpp:201]: There is no code to check `undef`, we just need to replace the function with return value here, but `nullptr` can't be as a new value.

I will try to fix this issue next week. and revert to the previous version first.

You can launch the CTS test 'dEQP-VK.api.copy_and_blit.copy_commands2.resolve_image.whole_array_image.4_bit',  the issue will be observed.

Thanks.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 25d62aab08324bc8bae4029c8ef04d1464363e7e
<details><summary>CTS tests (Failed: 0/286534)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.6, Gfx8</summary><ul>

<li>Passed: 36222/72653 (49.9%)</li>
<li>Failed: 0/72653 (0.0%)</li>
<li>Not Supported: 36431/72653 (50.1%)</li><li>Warnings: 0/72653 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi21</summary><ul>

<li>Passed: 40420/72700 (55.6%)</li>
<li>Failed: 0/72700 (0.0%)</li>
<li>Not Supported: 32280/72700 (44.4%)</li><li>Warnings: 0/72700 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 37135/72707 (51.1%)</li>
<li>Failed: 0/72707 (0.0%)</li>
<li>Not Supported: 35572/72707 (48.9%)</li><li>Warnings: 0/72707 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx9</summary><ul>

<li>Passed: 34834/68474 (50.9%)</li>
<li>Failed: 0/68474 (0.0%)</li>
<li>Not Supported: 33640/68474 (49.1%)</li><li>Warnings: 0/68474 (0.0%)</li>
</ul></details>


</ul></details>

**Review by nhaehnle — APPROVED** [review]
LGTM

**Comment by Flakebi** [comment]
This assertion is quite helpful to find non-matching resource layouts, thanks for adding it!
